lib/wasi.js

```diff
@@ -43,12 +44,37 @@ class WASI {
 
     // TODO(cjihrig): Validate preopen object schema.
 
-    // TODO(cjihrig): Temporarily expose these for development.
-    // eslint-disable-next-line no-undef
-    const memory = Buffer.alloc(200000);
-    this._memory = memory;
-    this._view = new DataView(memory.buffer);
-    this._wasi = new _WASI(args, envPairs, preopens, memory);
+    if (memory instanceof WebAssembly.Memory) {
+      memory = memory.buffer;
+    } else {
+      throw new ERR_INVALID_ARG_TYPE(
+        'options.memory', 'WebAssembly.Memory', memory);
+    }
+
+    const wrap = new _WASI(args, envPairs, preopens, memory);
+
+    for (const prop in wrap) {
+      wrap[prop] = wrap[prop].bind(wrap);
+    }
+
+    this.wasiImport = wrap;
+  }
+
+  static start(instance) {
```

Comment:
What happens if `instance.memory !== options.memory` passed to `new WASI`? Should this be considered an API hazard?

Reply:
Since `start()` is a static method, I don't think that should matter, right?

Reply:
Right I see, but then should there be some memory binding in this method instead?

Reply:
So, I'm assuming that `WASI.start()` is not the same thing as the API being discussed in https://github.com/nodejs/wasi/issues/3. I'm thinking of `WASI.start()` as more of a building block API, and https://github.com/nodejs/wasi/issues/3 as a higher level API.

With that assumption, consider the following example:

```js
'use strict';
const fs = require('fs');
const { WASI } = require('wasi');
const pathname = '/path/to/something.wasm';
const buffer = fs.readFileSync(pathname);
const memory = new WebAssembly.Memory({ initial: 1 });
const wasi = new WASI({ args: [], env: process.env, memory });

(async () => {
  const compiled = await WebAssembly.compile(buffer);
  const importObject = { wasi_unstable: wasi.wasiImport };
  const instance = await WebAssembly.instantiate(compiled, importObject);

  WASI.start(instance);
})();
```

In this case, `WASI.start()` just attempts to call the `_start()` or `__wasi_unstable_reactor_start()` export if one exists. By this point, `instance` should have everything bound properly, or `WebAssembly.instantiate()` will fail.

Reply:
I see. For commands that export their memories, a memory injection API would still likely be needed though.

Reply:
Agreed. I plan to have a `WASI.prototype.setMemory()` (or something along those lines) eventually.

Reply:
If setMemory can take a SharedArrayBuffer that could be awesome (although I am not sure if that is allowed)

Reply:
I just pushed up https://github.com/nodejs/wasi/pull/4/commits/2b0131c2f7e8961327ac81013927e1a659e8211a to this PR. It allows the `memory` passed to the `WASI` constructor to be a `SharedArrayBuffer` (or any other `ArrayBuffer`). Once `setMemory()` is added, it will also support SABs.

Reply:
This is really cool and useful thanks :]